src/destructors.md

```diff
@@ -37,57 +33,344 @@ impl Drop for ShowOnDrop {
-    partially_initialized.0 = ShowOnDrop("Partial tuple first");
+    // Tuple expression doesn't finish evaluating so operands drop in reverse order
+    (
+        ShowOnDrop("Outer tuple first"),
+        ShowOnDrop("Outer tuple second"),
+        (
+            ShowOnDrop("Inner tuple first"),
+            ShowOnDrop("Inner tuple second"),
+            break,
+        ),
+        ShowOnDrop("Never created"),
+    );
 }
 ```
 
+### Constant promotion
+
+Promotion of a value expression to a `'static` slot occurs when the expression
+could be written in a constant, borrowed, and dereferencing that borrow where
+the expression was originally written, without changing the runtime behavior.
+That is, the promoted expression can be evaluated at compile-time and the
+resulting value does not contain [interior mutability] or [destructors] (these
+properties are determined based on the value where possible, e.g. `&None`
+always has the type `&'static Option<_>`, as it contains nothing disallowed).
+
+### Temporary lifetime extension
+
+The temporary scopes for expressions in `let` statements are sometimes
+*extended* to the scope of the block containing the `let` statement. This is
+done when the usual temporary scope would be too small, based on certain
+syntactic rules. For example:
+
+```rust
+let x = &mut 0;
+// Usually a temporary would be dropped by now, but the temporary for `0` lives
+// to the end of the block.
+println!("{}", x);
+```
+
+If a borrow, dereference, field or tuple indexing expression has an extended
+temporary scope then so does its operand. If an indexing expression has an
+extended temporary scope then the indexed expression also has an extended
+temporary scope.
+
+#### Extending based on patterns
+
+An *extending pattern* is either
+
+* An [identifier pattern] that binds by reference or mutable reference.
+* A [struct][struct pattern], [tuple][tuple pattern], [tuple struct][tuple
+  struct pattern] or [slice][slice pattern] pattern where at least one of the
+  direct subpatterns is a extending pattern.
+
+So `ref x`, `V(ref x)` and `[ref x, y]` are all extending patterns, but `x`,
+`&ref x` and `&(ref x,)` are not.
+
+If the pattern in a `let` statement is an extending pattern then the temporary
+scope of the initializer expression is extended.
+
+#### Extending based on expressions
+
+For a let statement with an initializer, an *extending expression* is an
+expression which is one of the following:
+
+* The initializer expression.
+* The operand of a extending [borrow expression].
+* The operand(s) of an extending [array][array expression], [cast][cast
+  expression], [braced struct][struct expression] or [tuple][tuple expression]
+  expression.
+* The final expression of any extending [block expression][block expressions].
+
+So the borrow expressions in `&mut 0`, `(&1, &mut 2)` and `Some { 0: &mut 3 }`
+are all extending expressions, while the borrows in `&0 + &1` and
+`Some(&mut 0)` are not.
```

Comment:
`Some(&mut 0)` behaves this way because it is a function call?

Reply:
Right; can you add a note explaining why `Some(&mut 0)` is different from `Some { 0: &mut 3 }` since it may not be so clear for anyone reading this?